File: be/src/main/java/com/lavadero/service/ClientService.java
package com.lavadero.service;

import com.lavadero.dto.ClientRequest;
import com.lavadero.dto.ClientResponse;
import com.lavadero.dto.VehicleDto;
import com.lavadero.model.Client;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ClientService {
    ClientResponse createClient(ClientRequest request);
    ClientResponse updateClient(String id, ClientRequest request);
    void deleteClient(String id);
    ClientResponse getClientById(String id);
    Page<ClientResponse> getAllClients(Pageable pageable);
    List<ClientResponse> searchClientsByName(String searchTerm);
    ClientResponse addVehicle(String clientId, VehicleDto vehicle);
    ClientResponse removeVehicle(String clientId, String licensePlate);
}

File: fe/lavadero-fe/src/app/core/services/client.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Client, Vehicle } from '../models/models';

@Injectable({
  providedIn: 'root'
})
export class ClientService {
  private http = inject(HttpClient);

  getClients(page: number = 0, size: number = 20): Observable<any> {
    return this.http.get(`${environment.apiUrl}/clients?page=${page}&size=${size}`);
  }

  getClientById(id: string): Observable<Client> {
    return this.http.get<Client>(`${environment.apiUrl}/clients/${id}`);
  }

  createClient(client: Partial<Client>): Observable<Client> {
    return this.http.post<Client>(`${environment.apiUrl}/clients`, client);
  }

  updateClient(id: string, client: Partial<Client>): Observable<Client> {
    return this.http.put<Client>(`${environment.apiUrl}/clients/${id}`, client);
  }

  deleteClient(id: string): Observable<void> {
    return this.http.delete<void>(`${environment.apiUrl}/clients/${id}`);
  }

  searchClients(term: string): Observable<Client[]> {
    return this.http.get<Client[]>(`${environment.apiUrl}/clients/search?term=${term}`);
  }

  addVehicle(clientId: string, vehicle: Vehicle): Observable<Client> {
    return this.http.post<Client>(`${environment.apiUrl}/clients/${clientId}/vehicles`, vehicle);
  }

  removeVehicle(clientId: string, licensePlate: string): Observable<Client> {
    return this.http.delete<Client>(`${environment.apiUrl}/clients/${clientId}/vehicles/${licensePlate}`);
  }
}
